[DirectX][DXIL] Set DXIL Version in DXIL target triple based on shader model version

bharadwajy opened the pull request:
An earlier commit provided a way to decouple DXIL version from Shader Model version
by representing the DXIL version as `SubArch` in the DXIL Target Triple and adding 
corresponding valid DXIL Arch types.

This change constructs DXIL target triple with DXIL version that is deduced
from Shader Model version specified in the following scenarios:

1. When compilation target profile is specified: 
   For e.g., DXIL target triple `dxilv1.8-unknown-shader6.8-library` is constructed 
   when `-T lib_6_8` is specified. 
2. When DXIL target triple without DXIL version is specified:
    For e.g., DXIL target triple `dxilv1.8-pc-shadermodel6.8-library` is constructed
    when `-mtriple=dxil-pc-shadermodel6.8-library` is specified.

Note that this is an alternate / expanded implementation to that proposed in PR #89823.

PR #89823 implements functionality (1) above. However, it requires any DXIL target  
triple to explicitly include DXIL version anywhere DXIL target triple is expected (e.g., 
`dxilv1.8-pc-shadermodel6.8-library`) while a triple without DXIL version 
(e.g., `dxil-pc-shadermodel6.8-library`) is considered invalid. This functionality is
implemented as a change to `tryParseProfile()` and is included in this PR.

This PR adds the functionality (2) to eliminate the above requirement to explicitly 
specify DXIL version in the target triple thereby considering a triple without DXIL 
version ( e.g., `dxil-pc-shadermodel6.8-library`) to be valid. A triple with DXIL version 
is inferred based on the shader mode version specified. If no shader model version is 
specified, DXIL version is defaulted to 1.0. This functionality is implemented in the 
"Special case logic ..." section of `Triple::normalize()`. 
 
Updated relevant HLSL tests that check for target triple. 

Validated that Clang (`check-clang`) and LLVM (`check-llvm`) regression tests pass.

Commit:
Set DXIL Version in DXIL target triple based on shader model version
a) specified as compilation target profile or
b) specified as target triple string

Update relevant HLSL tests that check for target triple.

bob80905 commented on llvm/lib/TargetParser/Triple.cpp:
```diff
@@ -115,6 +115,30 @@ StringRef Triple::getArchName(ArchType Kind, SubArchType SubArch) {
+    case Triple::DXILSubArch_v1_8:
+      return "dxilv1.8";
+    default:
+      return "";
```
Would we want to llvm_unreachable or otherwise fail here, or does the caller handle this case well?

bharadwajy replied:
> Would we want to llvm_unreachable or otherwise fail here, or does the caller handle this case well?

The `default` case would be true for a call such as `getArchName(Triple::dxil, Triple::<NonDXILSubArch>)`,
where `NonDXILSubArch` is something other than `DXILSubArch_v1_[0..8]` (e.g., `MipsSubArch_r6`).

I considered 3 options to handle this case:

1. Report failure / crash with a message indicating incorrect DXIL `SubArch`:  not chosen as such behavior is not consistent with other `SuArch`s as pointed out [here](https://github.com/llvm/llvm-project/pull/89823/files#r1581314588) in the feedback for PR #89823.
2. Fall through to return the value of `getArchTypeName(Kind)` where `Kind` is `Triple::dxil` to return the string "dxil": not chosen since "dxil" does **not** represent the `ArchTypeName` for the `Arch/SubArch` combination of `Triple::dxil/Triple::<NonDXILSubArch>` (e.g., `Triple::dxil//Triple::MipsSubArch_r6`) and returning "dxil" would indicate that it is a valid `ArchTypeName`.
3. Return null string (""): chosen to indicate that there is no valid `ArchTypeName` for the `Arch/SubArch` combination of `Triple::dxil/Triple::<NonDXILSubArch>`(e.g., `Triple::dxil/Triple::MipsSubArch_r6`). The consumer of the Triple checks for validity of Arch to issue an error for a non-viable empty architecture name as I think it is a cleaner option compared to either reporting a failure or asserting here.

python3kgae commented on clang/lib/Basic/Targets.cpp:
```diff
@@ -760,7 +760,7 @@ using namespace clang::targets;
 TargetInfo::CreateTargetInfo(DiagnosticsEngine &Diags,
                              const std::shared_ptr<TargetOptions> &Opts) {
-  llvm::Triple Triple(Opts->Triple);
+  llvm::Triple Triple(llvm::Triple::normalize(Opts->Triple));
```
Why do we need to call llvm::Triple::normalize here?

python3kgae replied:
I see. It is trying to add subarch for dxil.
Could we allow Triple::NoSubArch for dxil?
Backend could do the job to translate NoSubArch to the version based on shader model.

bharadwajy replied:
> I see. It is trying to add subarch for dxil. Could we allow Triple::NoSubArch for dxil? 

`setArch(dxil, Triple::NoSubArch)` should work.

> Backend could do the job to translate NoSubArch to the version based on shader model.

One of the principles I tried to adhere to in the implementation of part (2) is to pass an appropriately constructed DXIL target triple string at creation-time of (DXIL) TargetMachine instance to maintain consistency between all the fields of the instance of DXIL `Triple` viz., Data, Arch, SubArch etc. 

Delegating or expecting the backend to do the same may lead to inconsistencies in the Triple instance. What benefit do you see in letting Backend perform the same steps that can be performed at creation time? Of course, if the backend needs to change the triple information for any reason, it is still free to do so - but I think it should expect to have the target triple information that reflects command-line or IR module specification of target triple.

python3kgae commented on llvm/lib/TargetParser/Triple.cpp:
```diff
@@ -115,6 +115,30 @@ StringRef Triple::getArchName(ArchType Kind, SubArchType SubArch) {
     break;
+  case Triple::dxil:
+    switch (SubArch) {
+    case Triple::NoSubArch:
```
Why default is dxil1.0?

bharadwajy replied:
> Why default is dxil1.0?

No specific reason other than choose the most basic version. 

Your comment [here](https://github.com/llvm/llvm-project/pull/89823#discussion_r1581541937) appeared to indicate a similar default choice for SPIRV (viz., spirv32 being equivalent to spirv32_v1.0).

I am open to suggestions for alternate default DXIL version choices.

bharadwajy commented on clang/lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -4675,7 +4676,7 @@ bool CompilerInvocation::CreateFromArgsImpl(
   InputKind DashX = Res.getFrontendOpts().DashX;
   ParseTargetArgs(Res.getTargetOpts(), Args, Diags);
-  llvm::Triple T(Res.getTargetOpts().Triple);
+  llvm::Triple T(llvm::Triple::normalize(Res.getTargetOpts().Triple));
```
> Looks like part of this change is to normalize `Res.getTargetOpts().Triple` in many places where it is used.
> 
> Would it be possible to arrange it so that this is normalized in one place earlier on, so all users of `getTargetOpts()` can assume that it is normalized already?

I believe the triple accessible using the expression `getTargetOpts().Triple` denotes the target options specified for the compilation. It seems appropriate to leave it unmodified (un-normalized) for reference or other purposes during compilation.

bogner replied:
IIUC `getTargetOpts().Triple` should *already* be normalized. The definition of the option includes a "Normalizer", like so:
```
def triple : Separate<["-"], "triple">,
  HelpText<"Specify target triple (e.g. i686-apple-darwin9)">,
  MarshallingInfoString<TargetOpts<"Triple">, "llvm::Triple::normalize(llvm::sys::getDefaultTargetTriple())">,
  AlwaysEmit, Normalizer<"normalizeTriple">;
```

So I think we shouldn't need the call to normalizer in CompilerInvocation at all, and we should look at the other places we're adding it to make sure they all make sense.

llvm-beanz commented on llvm/lib/TargetParser/Triple.cpp:
```diff
@@ -1200,6 +1224,27 @@ std::string Triple::normalize(StringRef Str) {
+    // DXIL version is not set for a non-specified OS string or one that does
+    // not starts with shadermodel.
+  }
+
```
We should add unit tests to test the normalization.

bharadwajy replied:
> We should add unit tests to test the normalization.

Additional unit tests added.

llvm-beanz reviewed (commented):
Given that @bogner had concerns about the other approach I think we should get him to review this before moving forward.

That said, other than some missing unit test coverage I think this approach looks fine.

bogner commented on llvm/lib/TargetParser/Triple.cpp:
```diff
@@ -1200,6 +1224,27 @@ std::string Triple::normalize(StringRef Str) {
+        // If shader model specified is other than 6.x, set DXIL Version to 1.0
+        Components[0] = DXILVerStr.append("0");
+      }
+    }
```
This doesn't look right. We've already parsed `Arch` and `OS` above, so I would expect this logic to look something like:
```c++
if (Arch == Triple::DXIL && SubArch == Triple::NoSubArch) {
  VersionTuple Ver = OS != Triple::UnknownOS ? parseVersionFromName(Components[2]) : VersionTuple();
  if (Ver)
    Components[0] = Components[0].append("1.").append(Ver.Minor);
  else (Ver)
    // default case...
}
```

bharadwajy replied:
The function `normalize` does not parse for `SubArch`. So, `SubArch` is not available  for checking.

An input DXIL triple string `Str` with or without version (e.g., `dxil-pc-shadermodel6.3` or `dxilv1.3-pc-shadermodel6.3` will set `Arch` to `Triple::dxil`. Implicit deduction and insertion of DXIL version should be done only is `Component[0] == "dxil"` and not otherwise. Hence the string equivalence check for "dxil" and not for `Arch == Triple::dxil`.

However, made code changes to leverage Arch and OS values already parsed as suggested. A couple of additional sanity checks are also added.

bogner commented on llvm/lib/CodeGen/CommandFlags.cpp:
```diff
@@ -744,7 +744,7 @@ void codegen::setFunctionAttributes(StringRef CPU, StringRef Features,
 codegen::createTargetMachineForTriple(StringRef TargetTriple,
                                       CodeGenOptLevel OptLevel) {
-  Triple TheTriple(TargetTriple);
+  Triple TheTriple(llvm::Triple::normalize(TargetTriple.str()));
```
It's a bit of a grey area because this function is only really used by testing tools but I don't think we should normalize here. If someone is running llvm-reduce or one of the fuzzers on a module with an invalid triple they may well be doing that intentionally.

bharadwajy replied:
> It's a bit of a grey area because this function is only really used by testing tools but I don't think we should normalize here. If someone is running llvm-reduce or one of the fuzzers on a module with an invalid triple they may well be doing that intentionally.

OK. Deleted the call to `normalize()` and restored the MIR test.

bogner commented on llvm/tools/opt/optdriver.cpp:
```diff
@@ -630,7 +630,7 @@ extern "C" int optMain(
   }
 
-  Triple ModuleTriple(M->getTargetTriple());
+  Triple ModuleTriple(Triple::normalize(M->getTargetTriple()));
```
Similarly to my concern about updating `createTargetMachineForTriple`, I don't think opt should messing with the triple it's given.

bharadwajy replied:
> Similarly to my concern about updating `createTargetMachineForTriple`, I don't think opt should messing with the triple it's given.

Deleted call to `normalize()`.

Commit:
Refine  normalization of DXIL triple string.
Add unit tests for DXIL triple normalization

Commit:
Delete redundant calls to normalize TargetOpts Triple as it is already normalized
as pointed out in PR feedback.

Commit:
Delete call normalize TargetOpts Triple in createTargetMachineForTriple
As a result any dxil triple specified in testing tools that invoke this
function is treated unmodified and intentional. Hence it is not modified
to include an implicit DXIL version - per PR feedback.

Restored change to MIR test file.

Commit:
Delete call normalize specified riple in opt driver
thereby using any dxil triple as specified.

bogner commented on llvm/lib/CodeGen/CommandFlags.cpp:
```diff
@@ -744,7 +744,7 @@ void codegen::setFunctionAttributes(StringRef CPU, StringRef Features,
 codegen::createTargetMachineForTriple(StringRef TargetTriple,
                                       CodeGenOptLevel OptLevel) {
-  Triple TheTriple(TargetTriple);
+  Triple TheTriple(TargetTriple.str());
```
This looks redundant - doesn't Triple's constructor do the conversion to `std::string` from Twine internally?

bharadwajy replied:
> This looks redundant - doesn't Triple's constructor do the conversion to `std::string` from Twine internally?

Thanks for flagging my incomplete restoration. Fixed.

bogner reviewed (approved):
Other than [my comment here](https://github.com/llvm/llvm-project/pull/90809/files#r1589771652) about what to do in getArchName this looks reasonable.

The string handling in `normalize` feels a bit unwieldy, but I don't have any obvious thoughts for how to simplify it.

Commit:
missed changes from PR feedback

python3kgae commented on llvm/unittests/TargetParser/TripleTest.cpp:
```diff
@@ -2454,4 +2454,19 @@ TEST(TripleTest, isArmMClass) {
+
+TEST(TripleTest, DXILNormaizeWithVersion) {
+  EXPECT_EQ("dxilv1.0-unknown-shadermodel6.0",
+            Triple::normalize("dxilv1.0--shadermodel6.0"));
```
What will Triple::normalize("dxilv1.0--shadermodel6.2") expect equal?
"dxilv1.2-unknown-shadermodel6.2"?

bharadwajy replied:
> What will Triple::normalize("dxilv1.0--shadermodel6.2") expect equal? "dxilv1.2-unknown-shadermodel6.2"?

No. Since DXIL version is specified in the triple, it is honored as is. So, it will be `dxilv1.0-unknown-shadermodel6.2`

python3kgae replied:
Do we need a rule for what is legal or not?

bharadwajy replied:
> So Triple::normalize("dxilv1.2--shadermodel6.0") will get "dxilv1.2-unknown-shadermodel6.0"?

Yes.

bharadwajy replied:
> Do we need a rule for what is legal or not?

I expect such validation to be part of later compiler passes.

Commit:
git-clang-format changes
